sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,519 @@
+
+    /// <inheritdoc/>
+    public override void Process(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        if (!_isLoggingEnabled)
+        {
+            ProcessNext(message, pipeline, currentIndex);
+            return;
+        }
+
+        ProcessSyncOrAsync(message, pipeline, currentIndex, async: false).EnsureCompleted();
+    }
+
+    /// <inheritdoc/>
+    public override async ValueTask ProcessAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex)
+    {
+        if (!_isLoggingEnabled)
+        {
+            await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            return;
+        }
+
+        await ProcessSyncOrAsync(message, pipeline, currentIndex, async: true).ConfigureAwait(false);
+    }
+
+    internal virtual void LogRequest()
+    {
+        if (!s_eventSource.IsEnabled())
+        {
+            return;
+        }
+    }
+
+    internal virtual void LogResponse()
+    {
+        if (!s_eventSource.IsEnabled())
+        {
+            return;
+        }
+    }
+
+    private async ValueTask ProcessSyncOrAsync(PipelineMessage message, IReadOnlyList<PipelinePolicy> pipeline, int currentIndex, bool async)
+    {
+        PipelineRequest request = message.Request;
+        string requestId = string.IsNullOrEmpty(_clientRequestIdHeaderName) ? string.Empty : GetRequestIdFromHeaders(request.Headers, _clientRequestIdHeaderName!);
+
+        s_eventSource.Request(request, requestId, _assemblyName, Sanitizer!);
+
+        Encoding? requestTextEncoding = null;
+
+        if (request.Headers.TryGetValue("Content-Type", out string? contentType))
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType!, out requestTextEncoding);
+        }
+
+        var logWrapper = new ContentEventSourceWrapper(s_eventSource, _logContent, _maxLength, message.CancellationToken);
+
+        await logWrapper.LogAsync(requestId, request.Content, requestTextEncoding, async).ConfigureAwait(false).EnsureCompleted(async);
+
+        var before = Stopwatch.GetTimestamp();
+
+        try
+        {
+            if (async)
+            {
+                await ProcessNextAsync(message, pipeline, currentIndex).ConfigureAwait(false);
+            }
+            else
+            {
+                ProcessNext(message, pipeline, currentIndex);
+            }
+        }
+        catch (Exception ex)
+        {
+            s_eventSource.ExceptionResponse(requestId, ex.ToString());
+            throw;
+        }
+
+        var after = Stopwatch.GetTimestamp();
+
+        PipelineResponse response = message.Response!;
+        bool isError = response.IsError;
+
+        response.Headers.TryGetValue("Content-Type", out string? responseContentType);
+        ContentTypeUtilities.TryGetTextEncoding(responseContentType!, out Encoding? responseTextEncoding);
+
+        bool wrapResponseContent = !message.BufferResponse &&
```

Comment:
It would be good to add a comment here to explain what each of these lines explicitly contributes to the decision whether or not to wrap the response content in a logging stream.  Why do we only do it if the log wrapper is enabled for errors, e.g.?  Why when the response is not buffered?  Why only when the content stream is not seekable?

We might be able to use some of the new APIs around buffered content to simplify this statement, but let's first get crisp on when we do this and why.